DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The following helper method stops an execution contect. \n",
+    "Before closing the provided `session`, it will persist the contained model to disk. \n",
+    "The model's filename is constructed from `checkpoint_path` and `global_step` and will be returned."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def stop_execution_context(execution_context, session, checkpoint_path=None, global_step=None):\n",
+    "    \n",
+    "    # If the model is not persisted, we'll return 'None'\n",
+    "    hibernation_path = None\n",
+    "    \n",
+    "    if checkpoint_path is not None and global_step is not None:\n",
+    "        # Saving session's model into checkpoint dir\n",
+    "        hibernation_path = execution_context[3].save(session, checkpoint_path, global_step=global_step)\n",
+    "        \n",
+    "    # Free all allocated resources of the session\n",
+    "    session.close()\n",
+    "    \n",
+    "    return hibernation_path"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now let's introduce the main routine for training and inference. \n",
+    "It takes a started execution context (given by `execution_context` and `session`), an optional `epoch` index and a flag wheather to calculate WER report data (`query_report`) or not.\n",
+    "Its main duty is to iterate over all batches and calculate the mean loss.\n",
+    "If an epoch is provided, it will also do back propagation - aka training.\n",
+    "It will return a batch set result tuple which consists of the mean loss and - if asked for report data - the mean accuracy and individual results tuple."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "def apply_batches(execution_context, session, epoch=-1, query_report=False):\n",
+    "    \n",
+    "    # An epoch of -1 means no train run\n",
+    "    do_training = epoch >= 0\n",
+    "    \n",
+    "    # Unpack variables\n",
+    "    if do_training:\n",
+    "        graph, data_set, tower_results, saver, apply_gradient_op, merged, writer = execution_context\n",
+    "    else:\n",
+    "        graph, data_set, tower_results, saver = execution_context\n",
+    "    results_params, _, avg_accuracy, avg_loss = tower_results\n",
+    "    \n",
+    "    batches_per_device = ceil(float(data_set.total_batches) / len(available_devices))\n",
     "\n",
-    "    # Apply gradients to modify the model\n",
-    "    apply_gradient_op = apply_gradients(optimizer, avg_tower_gradients)\n",
+    "    total_loss = 0.0\n",
+    "    params = [avg_loss]\n",
+    "    \n",
+    "    # Facilitate a train run\n",
+    "    if do_training:\n",
+    "        params.append([apply_gradient_op, merged])\n",
```

Comment:
Good point, missed that. I still think some restructuring here would help readability, but it might also require some shuffling elsewhere. It's too tangled for my liking - perhaps it would help if we defined some classes so that we can start using readable names rather than [0], [1], or having to extract things into separate variables? (i.e. we could say tower_results.params, or even execution_context.tower_results.params, or something along these lines - alternatively, getter functions)